x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,261 @@
+///
+#[allow(clippy::too_many_arguments)]
+pub fn swap_and_forward(
+    deps: DepsMut,
+    block_time: Timestamp,
+    contract_addr: Addr,
+    input_coin: Coin,
+    output_denom: String,
+    slippage: Slippage,
+    receiver: Addr,
+    next_memo: Option<String>,
+    failed_delivery: Option<Recovery>,
+) -> Result<Response, ContractError> {
+    deps.api.debug(&format!("executing swap and forward"));
+    let config = CONFIG.load(deps.storage)?;
+
+    // Message to swap tokens in the underlying swaprouter contract
+    let swap_msg = SwapRouterExecute::Swap {
+        input_coin: input_coin.clone(),
+        output_denom,
+        slippage,
+    };
+    let msg = wasm_execute(config.swap_contract, &swap_msg, vec![input_coin])?;
+
+    // Check that the received is valid and retrieve its channel
+    let (valid_channel, valid_receiver) = validate_receiver(deps.as_ref(), receiver)?;
+    // If there is a memo, check that it is valid
+    if let Some(memo) = &next_memo {
+        validate_memo(memo)?;
+    }
+
+    // Check that there isn't anything stored in SWAP_REPLY_STATES. If there is,
+    // it means that the contract is already waiting for a reply and should not
+    // override the stored state. This should only happen if a contract we call
+    // calls back to this one. This is likely a malicious attempt modify the
+    // contract's state before it has replied.
+    if SWAP_REPLY_STATES.may_load(deps.storage)?.is_some() {
+        return Err(ContractError::ContractLocked {
+            msg: "Already waiting for a reply".to_string(),
+        });
+    }
+    // Store information about the original message to be used in the reply
+    SWAP_REPLY_STATES.save(
+        deps.storage,
+        &SwapMsgReplyState {
+            swap_msg,
+            block_time,
+            contract_addr,
+            forward_to: ForwardTo {
+                channel: valid_channel,
+                receiver: valid_receiver,
+                next_memo,
+                failed_delivery,
+            },
+        },
+    )?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_success(msg, SWAP_REPLY_ID)))
+}
+
+pub fn handle_swap_reply(deps: DepsMut, msg: Reply) -> Result<Response, ContractError> {
+    deps.api.debug(&format!("handle_swap_reply"));
+    let swap_msg_state = SWAP_REPLY_STATES.load(deps.storage)?;
+    SWAP_REPLY_STATES.remove(deps.storage);
+
+    // If the swaprouter swap failed, return an error
+    let SubMsgResult::Ok(SubMsgResponse { data: Some(b), .. }) = msg.result else {
+        return Err(ContractError::FailedSwap {
+            msg: format!("No data"),
+        })
+    };
+
+    // Parse underlying response from the chain
+    let parsed =
+        cw_utils::parse_execute_response_data(&b).map_err(|e| ContractError::FailedSwap {
+            msg: format!("failed to parse: {e}"),
+        })?;
+    let swap_response: SwapResponse = from_binary(&parsed.data.unwrap())?;
```

Comment:
why don't we `.map_err(|e| ContractError::FailedSwap { .. })`  instead of `unwrap` here as well?

Reply:
parsed.data is an `Option`. I think we're garanteed to have data here, but it's probably better to do 

`let swap_response: SwapResponse = from_binary(&parsed.data.unwrap_or_default())?;`

anyway. That way if the data isn't there `from_binary` will fail with a nice error about not being able to decode SwapResponse

Reply:
Oh yeah unwrap_or_default make sense. Generally I think if it's not obvious that why it's guaranteed to work, I would prefer expect over unwrap.

Reply:
that makes sense! I changed it to unwrap_or_default(), but happy to change it to expect() if you have strong opinions